src/fetch-types.ts

```diff
@@ -1,31 +1,47 @@
-/// <reference lib="dom" />
+import type { Agent } from "http"
 import type { Assert } from "./type-utils"
+import type NodeFetchFn from "node-fetch"
 import type {
-  RequestInit as NodeRequestInit,
-  Response as NodeResponse,
+  RequestInfo as NodeFetchRequestInfo,
+  RequestInit as NodeFetchRequestInit,
+  Response as NodeFetchResponse,
 } from "node-fetch"
 
-type FetchFn = typeof fetch
-type FetchResponse = Awaited<ReturnType<FetchFn>>
-type RequestInfo = Parameters<FetchFn>[0]
-type RequestInit = NonNullable<Parameters<FetchFn>[1]>
-
+// The `Supported` types should be kept up to date in order to exactly match what we use in the client. This ensures maximal compatibility with other `fetch` implementations.
 export type SupportedRequestInfo = string
-type _assertSupportedInfoIsSubtype = Assert<RequestInfo, SupportedRequestInfo>
+// We can't assert against the browser or native Node fetch types without complicating the package structure (see #401), so perform a best effort against `node-fetch`, which we use by default.
+type _assertSupportedInfoIsSubtypeOfNodeFetch = Assert<
+  NodeFetchRequestInfo,
+  SupportedRequestInfo
+>
 
 export type SupportedRequestInit = {
-  agent?: NodeRequestInit["agent"]
-  body?: NonNullable<RequestInit["body"]> & NonNullable<NodeRequestInit["body"]>
-  headers?: NonNullable<RequestInit["headers"]> &
-    NonNullable<NodeRequestInit["headers"]>
-  method?: RequestInit["method"]
-  redirect?: RequestInit["redirect"]
+  agent?: Agent
+  body?: string
```

Comment:
Is there any harm to also including the `node-fetch` body types (i.e. `NodeRequestInit["body"]`) here too? I'm concerned that there might be folks using `Blob` or stream types via node here that this breaks

Reply:
The short answer is no.

The long answer is (my apologies if I’m explaining stuff you already know): TypeScript uses duck typing, which means that as long as the user-provided `fetch` implementation accepts `body: string`, it doesn’t matter what else their fetch implementation accepts. We’re just specifying the minimal contract of things we need to use, which in our case is just `string | undefined` since in `Client.ts`, the only place `fetch` is used, we pass in the result of `JSON.stringify()` or `undefined`. For example, tRPC accept custom `fetch` implementations but [only type a subset of `body`](https://github.com/trpc/trpc/blob/8234a83b48a927337f956ea665231c5c1575e127/packages/client/src/internals/types.ts#L55), which corresponds with what they actually use internally.

However, in practice, extending our contract wouldn’t hurt. Right now we type `body: NonNullable<RequestInit["body"]> & NonNullable<NodeRequestInit["body"]>`, which uses an intersection over the DOM and `node-fetch` types. This computes to something like `body: Blob | URLSearchParams | FormData | string` (I’m hazy on this, but the point is that, as an intersection, it removes anything that’s not the same type between implementations, such as `NodeJS.ReadableStream`). In this PR, taking the intersection over `string & NodeRequestInit["body"]` would compute to `string`, which is kind of pointless (but doing `string | NodeRequestInit["body"]` would cause all manner of incompatibilities because we’d require every client to accept `NodeJS.ReadableStream`, to give an example).

So I think it wouldn’t make sense to add this.

Reply:
Thanks for the explanation! I had missed that we don't allow the user to pass through a `body`, and always pass through the result of `JSON.stringify()` -- so makes sense why keeping `string` is perfectly fine.